### PR TITLE
Use whip instead of macros for HIP/CUDA compatibility

### PR DESCRIPTION
**Pull request by msimberg**
Companion of https://github.com/pika-org/pika/pull/423.

The translation is quite straightforward. I so far put `whip::impl::check_error` (which throws an exception on non-success) into `impl` in whip. ~I should probably move that out of `impl` and then it could be used to replace the few places where I now explicitly check the status and throw a `whip::exception`.~ Done.

Any kind of comments most welcome.

### CODE REVIEW

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try-

**Comment by msimberg** [comment]
bors try-

**Comment by msimberg** [comment]
bors try-

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try-

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
This is ready for reviews, but need not be merged even if approved since there's no rush. It would probably be a good idea to wait until the HIP CI is working again for this PR.

**Comment by msimberg** [comment]
bors try

File: ci/.gitlab-ci.yml
Comment (msimberg) [inline ci/.gitlab-ci.yml:47]: Revert before merging.

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
Assuming CI is now happy this would be ready to go from my side. Reviews welcome.

**Review by rasolca — APPROVED** [review]

**Comment by msimberg** [comment]
bors try

**Pull request merged**
